dev/PipsControl/PipsControl.cpp

```diff
@@ -0,0 +1,369 @@
+    const auto templateSettings = winrt::make<PipsControlTemplateSettings>();
+    templateSettings.SetValue(PipsControlTemplateSettings::s_VerticalPipsItemsProperty, m_verticalPipsElements);
+    SetValue(s_TemplateSettingsProperty, templateSettings);
+
+    SetDefaultStyleKey(this);
+
+}
+
+PipsControl::~PipsControl() {
+    m_nextPageButtonClickRevoker.revoke();
+    m_previousPageButtonClickRevoker.revoke();
+    m_verticalPipsElementPreparedRevoker.revoke();
+    m_rootGridPointerEnteredRevoker.revoke();
+    m_rootGridPointerExitedRevoker.revoke();
+    m_rootGridKeyDownRevoker.revoke();
+}
+
+void PipsControl::OnApplyTemplate()
+{
+    winrt::AutomationProperties::SetName(*this, ResourceAccessor::GetLocalizedStringResource(SR_PipsControlNameText));
+
+    m_previousPageButtonClickRevoker.revoke();
+    [this](const winrt::Button button) {
+        if (button) {
+            winrt::AutomationProperties::SetName(button, ResourceAccessor::GetLocalizedStringResource(SR_PipsControlPreviousPageButtonText));
+            m_previousPageButtonClickRevoker = button.Click(winrt::auto_revoke, { this, &PipsControl::OnPreviousButtonClicked });
+        }
+    }(GetTemplateChildT<winrt::Button>(c_previousPageButtonName, *this));
+
+    m_nextPageButtonClickRevoker.revoke();
+    [this](const winrt::Button button) {
+        if (button) {
+            winrt::AutomationProperties::SetName(button, ResourceAccessor::GetLocalizedStringResource(SR_PipsControlNextPageButtonText));
+            m_nextPageButtonClickRevoker = button.Click(winrt::auto_revoke, { this, &PipsControl::OnNextButtonClicked });
+        }
+    }(GetTemplateChildT<winrt::Button>(c_nextPageButtonName, *this));
+
+    m_verticalPipsElementPreparedRevoker.revoke();
+    [this](const winrt::ItemsRepeater repeater) {
+        m_verticalPipsRepeater.set(repeater);
+        if (repeater) {
+            m_verticalPipsElementPreparedRevoker = repeater.ElementPrepared(winrt::auto_revoke, { this, &PipsControl::OnElementPrepared });
+        }
+    }(GetTemplateChildT<winrt::ItemsRepeater>(c_PipsControlPipsRepeaterName, *this));
+
+
+    m_verticalPipsScrollViewer.set(GetTemplateChildT<winrt::FxScrollViewer>(c_PipsControlScrollViewerName, *this));
+
+    m_rootGridPointerEnteredRevoker.revoke();
+    m_rootGridPointerExitedRevoker.revoke();
+
+    [this](const winrt::Grid grid) {
+        if (grid) {
+            m_rootGridPointerEnteredRevoker = grid.PointerEntered(winrt::auto_revoke, { this, &PipsControl::OnPipsControlPointerEntered });
+            m_rootGridPointerExitedRevoker = grid.PointerExited(winrt::auto_revoke, { this, &PipsControl::OnPipsControlPointerExited });
+            m_rootGridKeyDownRevoker = grid.KeyDown(winrt::auto_revoke, { this, &PipsControl::OnRootGridKeyDown });
+
+        }
+    }(GetTemplateChildT<winrt::Grid>(c_rootGridName, *this));
+
+    OnNumberOfPagesChanged(0);
```

Comment:
>OnNumberOfPagesChanged [](start = 4, length = 22)

I think this breaks the control if the current NumberOfPages isn't 0... for instance if we apply a new template to the control that already has a NumberOfPages set.  I think instead of taking a value as a parameter these functions should look up the value themselves using the public getter (NumberOfPage())

Reply:
The numbers that are passed in are the old values - the ones that've been set before we hit these lines.
I can substitute them with the **m_lastSelectedPageIndex** etc., but I'm not sure if we need to do this
I might be not understanding something - could you please clarify if I'm mistaken?